tokenutil/token_util.go

```diff
@@ -0,0 +1,117 @@
+package tokenutil
+
+import (
+	"encoding/json"
+	"fmt"
+	"io/ioutil"
+	"net/http"
+	"net/url"
+	"os"
+	"strings"
+
+	"github.com/Azure/go-autorest/autorest"
+	"github.com/Azure/go-autorest/autorest/adal"
+	"github.com/Azure/go-autorest/autorest/azure/auth"
+	"github.com/pkg/errors"
+)
+
+const (
+	// environment variable to override the default msi endpoint
+	envMsiEndpoint = "MSI_ENDPOINT"
+)
+
+// RegistryRefreshToken is the response body from ACR exchange API
+type RegistryRefreshToken struct {
+	RefreshToken string `json:"refresh_token"`
+}
+
+// GetRegistryRefreshToken gets ARM token, and makes a POST request to registry/v2 endpoint to
+// exchange it for Registry token. Steps mentioned in detail below
+// Authentication: https://github.com/Azure/acr/blob/master/docs/AAD-OAuth.md#authenticating-to-a-registry-with-azure-cli
+// Exchange: https://github.com/Azure/acr/blob/master/docs/AAD-OAuth.md#calling-post-oauth2exchange-to-get-an-acr-refresh-token
+// Note, we don't need to do token challenge part.
+func GetRegistryRefreshToken(registry, resourceID, clientID string) (string, error) {
```

Comment:
AccessToken

Reply:
@northtyphoon as I mentioned yesterday. ACR gives back a `refresh_token`. This is similar code as in the CLI